Beistellung_Implementierung/system/Coding/src/app/StateHandler.c

```diff
@@ -0,0 +1,267 @@
+/***************************************************************************************************
+  (c) NewTec GmbH 2024   -   www.newtec.de
+***************************************************************************************************/
+/**
+ * @file       StateHandler.c
+ *
+ *    Decides which states are called next depending on the returned events  
+ */
+/**************************************************************************************************/
+
+/* INCLUDES ***************************************************************************************/
+#include "StateHandler.h"
+#include "EventEnum.h"
+
+#include "app/State/InitializeMcu.h"
+#include "app/State/CalibrateLineSensors.h"
+#include "app/State/ReadyToDrive.h"
+#include "app/State/DriveToStart.h"
+#include "app/State/DriveToFinish.h"
+#include "app/State/DriveOverGap.h"
+#include "app/State/DisplayCountdown.h"
+#include "app/State/DisplayLapTime.h"
+#include "app/State/ErrorHandlerState.h"
+#include "app/State/SetParameters.h"
+/* CONSTANTS **************************************************************************************/
+
+/* MACROS *****************************************************************************************/
+
+/* TYPES ******************************************************************************************/
+
+/* PROTOTYPES *************************************************************************************/
+/** The next state is decided on the basis of the transferred parameter 
+ * @param[in] gCurrentEvent Event that occurred in Drive to Start
+*/
+static void driveToStartEvent(EventEnum gCurrentEvent);
+
+/** The next state is decided on the basis of the transferred parameter 
+ * @param[in] gCurrentEvent Event that occurred in Drive to Finish
+*/
+static void driveToFinishEvent(EventEnum gCurrentEvent);
+
+/** The next state is decided on the basis of the transferred parameter 
+ * @param[in] gCurrentEvent Event that occurred in Drive Over Gap
+*/
+static void driveOverGapEvent(EventEnum gCurrentEvent);
+
+/** The next state is decided on the basis of the transferred parameter 
+ * @param[in] gCurrentEvent Event that occurred in Count Down Event
+*/
+static void displayCountDownEvent(EventEnum gCurrentEvent);
+
+/** The next state is decided on the basis of the transferred parameter 
+ * @param[in] gCurrentEvent Event that occurred in Ready to Drive
+*/
+static void readyToDriveEvent(EventEnum gCurrentEvent);
+
+/** The next state is decided on the basis of the transferred parameter 
+ * @param[in] gCurrentEvent Event that occurred in Calibrate Line Sensors
+*/
+static void calibrateLineSensorsEvent(EventEnum gCurrentEvent);
+/* VARIABLES **************************************************************************************/
+static EventEnum gCurrentEvent = NO_EVENT_HAS_HAPPEND; 
+static StateEnum gCurrentState = STATE_INIZALIZATION_MCU; /**< First state when the MCU is started */
+static SoftTimer* gLapTimer;
```

Comment:
as a tip, having an uninitialized pointer is dangerous.
You could initialize it as NULL and then check that it is not NULL before using it in a function.
Accidental writes on undefined pointer adresses cause unpredictable behaviour and are nearly impossible to find with debugging

Reply:
If we check for an uninitialized pointer we need to do something if the check results true. This means we need proper error handling. We do not have proper error handling at other locations in our code, because this was not intended by our approved design. To really fix this issue, we should implement proper error handling not just here but everywhere. This would also lead to lots of changes to the design. We will discuss this and maybe change it later...

Reply:
(We already made a lot of changes to the code of the implementation phase for a running final version. Therefore changes won't be adresses directly in this old version but in the new one.)

Now corrected this specific issue in the new version.

The proper error handling of the whole project cannot be fixed in the given time with the given code. The reason is that this is not only a problem of our application code but also of any other module outside the application layer. Every other layer is affected in many locations: hal, os and service. Only adjusting the application layer wouldn't be enough. If we really want to fix the problem, we have to implement proper error handling in all layers. But that is just way too much. So we just fixed the problem at hand and added an error case for the mentioned problem.

Reply:
If you want we can discuss possibilities for error handling in such cases on friday.

You do not necessarily need a complicated error handling, just some way to notify what happened.  If the error is not handled at all it could cause problems at a different location and it will just be harder to find the source and fix it.